### ISSUE TITLE
[Bug] Signed Distance produces erroneous result

### ISSUE THREAD

**Issue by Marviel** [mikedh/trimesh#85]
MESH: https://drive.google.com/file/d/0B18jNduc8kjlZW0zWHl4aEx3cXc/view?usp=sharing

TEST CASE:
```
mesh = trimesh.load_mesh("PROVIDEDMESH.stl")
points = [[4.5, 0, -23]] #outside the mesh
ret = trimesh.proximity.signed_distance(mesh, points)
assert np.all(ret < 0)
```

Despite appearances, this is not a duplicate of #73

**Comment by mikedh** [comment]
Thanks for the report! Man, that is tricky... it actually is hitting an edge 3 times despite being outside the mesh. Not sure what to do in the general case here. 

The cheesy thing to do which would fix this example at least would be to do an AABB test on the points before the ray test. That might be a good speedup regardless and I might add it, but that doesn't solve this case if the point is inside the AABB (like if there was an additional non-convex feature surrounding the point). 

I could make every ray test bidirectional, and if the two directions disagree do a special case. Maybe that + the AABB would be enough to get most of these cases without having to do awful things.

**Comment by Marviel** [comment]
No prob! Sorry to be difficult! XD

Yeah I figured that duplicating the raytest with a slightly modified ray, or a ray in the other direction, wouldn't be a bad choice. But also seems only to be necessary in certain cases.

One thing that would work for my particular use case (and likely be simpler to code up) would be to have the option of providing a ray direction to the signed distance function. In this way, if the results for a few points looked suspicious, then I could re run those with a slightly different ray direction vector.

That, plus the option to go ahead and do X number of ray tests per point like you were saying, and then the AABB test.

I'm not quite sure what the best choice is... This may be dumb but is there any easy way to tell when rays hit the edge? I can't seem to find it in the code... if so, would there be a way to note all different points that hit an edge, and re-run their ray tests?

**Comment by mikedh** [comment]
No worries, thanks for actually checking these edge cases! I added some [logic last night](https://github.com/mikedh/trimesh/blob/master/trimesh/ray/ray_util.py#L9-L85), which survives all the edge cases you've discovered, let me know if you can make that break 😅. I also added your model + check to the unit tests.  

The new logic is basically doing:

* Cast rays bidirectionally (along a random vector, rather than the Z axis to avoid models with features along Z). 
* If the hit count agrees in both directions, then we use that result. 
* If one of the directions hits nothing, we declare the point in free space even if the counts disagree. I haven't seen an edge case where we miss hits yet, but this step is pretty arbitrary and may need revisiting later
* If the two directions disagree and the point is not in free space, we try a contains test again for those points (but only once to avoid infinite recursion) along a random vector

**Comment by Marviel** [comment]
Awesome @mikedh !

Just a quick q -- does this mean that the raycasting will take roughly twice as long?

**Comment by mikedh** [comment]
Yeah, most of the time is probably in the ray query. I added an AABB cull in https://github.com/mikedh/trimesh/commit/e995a2139353712dcd47b5946f774a63e6440397 which should help speed up some distributions of points.

**Comment by Marviel** [comment]
Ah I see, makes sense.

The distributions I'm looking at tend to mostly be within the AABB, so this impacts speed significantly.

Would it be possible to (for the Raycasting part at least) keep the original approach, with the single ray per point -- and then any points whose rays hit edges, run again with another ray? This seems like it would take less time potentially.

Not sure how hard edge-hit detection is though, as I haven't looked into it.

**Comment by mikedh** [comment]
I'm not super confident in the on-edge/on-vertex detection being robust to weirdness. By the way have you installed pyembree, its roughly 55x faster :stuck_out_tongue: 

That being said the easy way to implement it would be to do an ray.intersects_location and then do a triangles.points_to_barycentric, and any barycentric coordinate with a zero term (b1,b2,(1.0-b1-b2)) is either on-edge (one zero) or on-vertex (two zeros). And then you could do bidirectional agree- logic (like I have implemented now for every point) to resolve those. 

Not sure if in the end it would work out to be faster but is probably worth trying.

**Comment by mikedh** [comment]
Cool since the results aren't wrong just slower I'm going to close this one. Feel free to open a performance issue, especially if you find a particularly egregious example 😄
